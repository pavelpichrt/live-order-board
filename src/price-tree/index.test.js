const PriceTree = require('./');

// Initial stubs don't contain any duplicate prices
const priceQuantitiesStubs = require('./__stubs__/price-quantities');

describe('PriceTree', () => {
	let priceTree;

	describe('sorted in ASCENDING order', () => {
		beforeEach(() => {
			priceTree = new PriceTree(PriceTree.ASC);
		});

		describe('getOrderedValues', () => {
			const existingPrice = priceQuantitiesStubs[0].price;
			const currentQuantity = priceQuantitiesStubs[0].quantity;
			const newOrderQuantity = 1.5;
			const samePriceOrderStub = {
				price: existingPrice,
				quantity: newOrderQuantity,
			};

			beforeEach(() => {
				priceQuantitiesStubs.forEach(({ price, quantity }) => {
					priceTree.addQuantity(price, quantity);
				});
			});

			describe('addQuantity', () => {
				test('aggregates quantities at DIFFERENT prices correctly', () => {
					expect(priceTree.getOrderedValues()).toHaveLength(priceQuantitiesStubs.length);
					expect(priceTree.getOrderedValues()).toMatchSnapshot();
				});

				test('aggregates quantities correctly with multiple quantities at the SAME price', () => {
					priceQuantitiesStubs.slice(0, 2).forEach(({ price, quantity }) => {
						priceTree.addQuantity(price, quantity);
					});

					expect(priceTree.getOrderedValues()).toHaveLength(priceQuantitiesStubs.length);
					expect(priceTree.getOrderedValues()).toMatchSnapshot();
				});
			});

			describe('removeQuantity', () => {
				test('removes the price when the remaining quantity is 0', () => {
					const priceStub = priceQuantitiesStubs[1];

					priceTree.removeQuantity(priceStub.price, priceStub.quantity);

					expect(priceTree.getOrderedValues()).toHaveLength(priceQuantitiesStubs.length - 1);
					expect(priceTree.getOrderedValues()).toMatchSnapshot();
				});

				describe('when the remaining quantity is greater than 0', () => {
					beforeEach(() => {
						// first add some quantities at duplicate price points
						priceQuantitiesStubs.slice(0, 2).forEach(({ price, quantity }) => {
							priceTree.addQuantity(price, quantity);
						});

						const priceStub = priceQuantitiesStubs[1];

						priceTree.removeQuantity(priceStub.price, priceStub.quantity);
					});

					test('decreases the quantity for a price correctly', () => {
						expect(priceTree.getOrderedValues()).toMatchSnapshot();
					});

					test('does NOT remove the price from the tree', () => {
						expect(priceTree.getOrderedValues()).toHaveLength(priceQuantitiesStubs.length);
					});
				});
			});
		});
	});

	describe('sorted in DESCENDING order', () => {
		beforeEach(() => {
			priceTree = new PriceTree(PriceTree.DESC);
		});

		describe('getOrderedValues', () => {
			const existingPrice = priceQuantitiesStubs[0].price;
			const currentQuantity = priceQuantitiesStubs[0].quantity;
			const newOrderQuantity = 1.5;
			const samePriceOrderStub = {
				price: existingPrice,
				quantity: newOrderQuantity,
			};

			beforeEach(() => {
				priceQuantitiesStubs.forEach(({ price, quantity }) => {
					priceTree.addQuantity(price, quantity);
				});
			});

			describe('addQuantity', () => {
				test('aggregates quantities at DIFFERENT prices correctly', () => {
					expect(priceTree.getOrderedValues()).toHaveLength(priceQuantitiesStubs.length);
					expect(priceTree.getOrderedValues()).toMatchSnapshot();
				});

				test('aggregates quantities correctly with multiple quantities at the SAME price', () => {
					priceQuantitiesStubs.slice(0, 2).forEach(({ price, quantity }) => {
						priceTree.addQuantity(price, quantity);
					});

					expect(priceTree.getOrderedValues()).toHaveLength(priceQuantitiesStubs.length);
					expect(priceTree.getOrderedValues()).toMatchSnapshot();
				});
			});

			describe('removeQuantity', () => {
				test('removes the price when the remaining quantity is 0', () => {
					const priceStub = priceQuantitiesStubs[1];

					priceTree.removeQuantity(priceStub.price, priceStub.quantity);

					expect(priceTree.getOrderedValues()).toHaveLength(priceQuantitiesStubs.length - 1);
					expect(priceTree.getOrderedValues()).toMatchSnapshot();
				});

				describe('when the remaining quantity is greater than 0', () => {
					beforeEach(() => {
						// first add some quantities at duplicate price points
						priceQuantitiesStubs.slice(0, 2).forEach(({ price, quantity }) => {
							priceTree.addQuantity(price, quantity);
						});

						const priceStub = priceQuantitiesStubs[1];

						priceTree.removeQuantity(priceStub.price, priceStub.quantity);
					});

					test('decreases the quantity for a price correctly', () => {
						expect(priceTree.getOrderedValues()).toMatchSnapshot();
					});

					test('does NOT remove the price from the tree', () => {
						expect(priceTree.getOrderedValues()).toHaveLength(priceQuantitiesStubs.length);
					});
				});
			});
		});
	});
});
