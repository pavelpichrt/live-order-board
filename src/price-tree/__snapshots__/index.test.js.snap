// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceTree sorted in ASCENDING order getOrderedValues addQuantity aggregates quantities at DIFFERENT prices correctly 1`] = `
Array [
  Object {
    "price": 55.8,
    "quantity": 13.5,
  },
  Object {
    "price": 69,
    "quantity": 6.4,
  },
  Object {
    "price": 76.1,
    "quantity": 5.88,
  },
  Object {
    "price": 323.77,
    "quantity": 4.2,
  },
]
`;

exports[`PriceTree sorted in ASCENDING order getOrderedValues addQuantity aggregates quantities correctly with multiple quantities at the SAME price 1`] = `
Array [
  Object {
    "price": 55.8,
    "quantity": 27,
  },
  Object {
    "price": 69,
    "quantity": 6.4,
  },
  Object {
    "price": 76.1,
    "quantity": 5.88,
  },
  Object {
    "price": 323.77,
    "quantity": 8.4,
  },
]
`;

exports[`PriceTree sorted in ASCENDING order getOrderedValues removeQuantity removes the price when the remaining quantity is 0 1`] = `
Array [
  Object {
    "price": 69,
    "quantity": 6.4,
  },
  Object {
    "price": 76.1,
    "quantity": 5.88,
  },
  Object {
    "price": 323.77,
    "quantity": 4.2,
  },
]
`;

exports[`PriceTree sorted in ASCENDING order getOrderedValues removeQuantity when the remaining quantity is greater than 0 decreases the quantity for a price correctly 1`] = `
Array [
  Object {
    "price": 55.8,
    "quantity": 13.5,
  },
  Object {
    "price": 69,
    "quantity": 6.4,
  },
  Object {
    "price": 76.1,
    "quantity": 5.88,
  },
  Object {
    "price": 323.77,
    "quantity": 8.4,
  },
]
`;

exports[`PriceTree sorted in DESCENDING order getOrderedValues addQuantity aggregates quantities at DIFFERENT prices correctly 1`] = `
Array [
  Object {
    "price": 323.77,
    "quantity": 4.2,
  },
  Object {
    "price": 76.1,
    "quantity": 5.88,
  },
  Object {
    "price": 69,
    "quantity": 6.4,
  },
  Object {
    "price": 55.8,
    "quantity": 13.5,
  },
]
`;

exports[`PriceTree sorted in DESCENDING order getOrderedValues addQuantity aggregates quantities correctly with multiple quantities at the SAME price 1`] = `
Array [
  Object {
    "price": 323.77,
    "quantity": 8.4,
  },
  Object {
    "price": 76.1,
    "quantity": 5.88,
  },
  Object {
    "price": 69,
    "quantity": 6.4,
  },
  Object {
    "price": 55.8,
    "quantity": 27,
  },
]
`;

exports[`PriceTree sorted in DESCENDING order getOrderedValues removeQuantity removes the price when the remaining quantity is 0 1`] = `
Array [
  Object {
    "price": 323.77,
    "quantity": 4.2,
  },
  Object {
    "price": 76.1,
    "quantity": 5.88,
  },
  Object {
    "price": 69,
    "quantity": 6.4,
  },
]
`;

exports[`PriceTree sorted in DESCENDING order getOrderedValues removeQuantity when the remaining quantity is greater than 0 decreases the quantity for a price correctly 1`] = `
Array [
  Object {
    "price": 323.77,
    "quantity": 8.4,
  },
  Object {
    "price": 76.1,
    "quantity": 5.88,
  },
  Object {
    "price": 69,
    "quantity": 6.4,
  },
  Object {
    "price": 55.8,
    "quantity": 13.5,
  },
]
`;
